[elasticsearchexporter] Direct serialization without objmodel in OTel mode

felixbarny opened the pull request:
Directly serializes pdata to JSON in OTel mode
* Improved performance as no `objmodel.Document` needs to be created first
* Fixes issue discovered in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37021 where map bodies with dotted field names are de-dotted

felixbarny committed:
Serialize logs directly to JSON in OTel mode

(skipping objmodel)

felixbarny committed:
Serialize spans and span events without objmodel

felixbarny committed:
Serialize metrics without objmodel

felixbarny committed:
Add changelog

felixbarny committed:
Merge remote-tracking branch 'origin/main' into es-direct-serialization

felixbarny committed:
goporto

felixbarny committed:
Fix linting issues

felixbarny committed:
Merge remote-tracking branch 'origin/main' into es-direct-serialization

felixbarny commented:
Benchmark results:

TL;DR: the throughput is almost 2x for metrics and over 2x for logs and traces. The allocated bytes/op are reduced by over 70% across the board

```
                                      │   old.txt   │               new.txt                │
                                      │   sec/op    │    sec/op     vs base                │
Exporter/logs/otel/small_batch-10       79.16µ ± 1%   37.09µ ±  4%  -53.15% (p=0.000 n=10)
Exporter/logs/otel/medium_batch-10      757.0µ ± 2%   296.9µ ±  6%  -60.78% (p=0.000 n=10)
Exporter/logs/otel/large_batch-10       7.392m ± 1%   2.793m ±  1%  -62.22% (p=0.000 n=10)
Exporter/logs/otel/xlarge_batch-10      70.50m ± 1%   27.57m ±  0%  -60.89% (p=0.000 n=10)
Exporter/metrics/otel/small_batch-10    414.8µ ± 1%   211.0µ ±  1%  -49.13% (p=0.000 n=10)
Exporter/metrics/otel/medium_batch-10   3.960m ± 0%   2.027m ±  1%  -48.82% (p=0.000 n=10)
Exporter/metrics/otel/large_batch-10    39.97m ± 1%   21.37m ±  0%  -46.54% (p=0.000 n=10)
Exporter/metrics/otel/xlarge_batch-10   421.3m ± 1%   228.2m ± 12%  -45.85% (p=0.000 n=10)
Exporter/traces/otel/small_batch-10     79.64µ ± 0%   31.92µ ±  1%  -59.92% (p=0.000 n=10)
Exporter/traces/otel/medium_batch-10    765.5µ ± 1%   272.8µ ±  1%  -64.36% (p=0.000 n=10)
Exporter/traces/otel/large_batch-10     7.341m ± 1%   2.608m ±  1%  -64.48% (p=0.000 n=10)
Exporter/traces/otel/xlarge_batch-10    71.74m ± 1%   25.73m ±  2%  -64.13% (p=0.000 n=10)
geomean                                 4.171m        1.783m        -57.25%

                                      │   old.txt   │                new.txt                │
                                      │  events/s   │   events/s    vs base                 │
Exporter/logs/otel/small_batch-10       126.3k ± 1%   269.6k ±  4%  +113.43% (p=0.000 n=10)
Exporter/logs/otel/medium_batch-10      132.1k ± 2%   336.8k ±  6%  +154.97% (p=0.000 n=10)
Exporter/logs/otel/large_batch-10       135.3k ± 1%   358.1k ±  1%  +164.69% (p=0.000 n=10)
Exporter/logs/otel/xlarge_batch-10      141.8k ± 1%   362.7k ±  0%  +155.70% (p=0.000 n=10)
Exporter/metrics/otel/small_batch-10    168.8k ± 1%   331.8k ±  1%   +96.59% (p=0.000 n=10)
Exporter/metrics/otel/medium_batch-10   176.7k ± 0%   345.3k ±  1%   +95.39% (p=0.000 n=10)
Exporter/metrics/otel/large_batch-10    175.1k ± 1%   327.6k ±  0%   +87.05% (p=0.000 n=10)
Exporter/metrics/otel/xlarge_batch-10   166.1k ± 1%   306.8k ± 11%   +84.66% (p=0.000 n=10)
Exporter/traces/otel/small_batch-10     125.6k ± 0%   313.3k ±  1%  +149.52% (p=0.000 n=10)
Exporter/traces/otel/medium_batch-10    130.6k ± 1%   366.5k ±  1%  +180.58% (p=0.000 n=10)
Exporter/traces/otel/large_batch-10     136.2k ± 0%   383.5k ±  1%  +181.50% (p=0.000 n=10)
Exporter/traces/otel/xlarge_batch-10    139.4k ± 1%   388.6k ±  2%  +178.79% (p=0.000 n=10)
geomean                                 145.0k        339.3k        +133.93%

                                      │   old.txt    │               new.txt                │
                                      │     B/op     │     B/op      vs base                │
Exporter/logs/otel/small_batch-10       80.58Ki ± 0%   16.98Ki ± 1%  -78.93% (p=0.000 n=10)
Exporter/logs/otel/medium_batch-10      793.1Ki ± 0%   158.4Ki ± 0%  -80.02% (p=0.000 n=10)
Exporter/logs/otel/large_batch-10       7.727Mi ± 0%   1.527Mi ± 0%  -80.24% (p=0.000 n=10)
Exporter/logs/otel/xlarge_batch-10      77.16Mi ± 0%   15.20Mi ± 0%  -80.30% (p=0.000 n=10)
Exporter/metrics/otel/small_batch-10    403.9Ki ± 0%   112.1Ki ± 0%  -72.25% (p=0.000 n=10)
Exporter/metrics/otel/medium_batch-10   3.926Mi ± 0%   1.070Mi ± 0%  -72.75% (p=0.000 n=10)
Exporter/metrics/otel/large_batch-10    39.51Mi ± 0%   11.08Mi ± 0%  -71.97% (p=0.000 n=10)
Exporter/metrics/otel/xlarge_batch-10   390.3Mi ± 0%   130.3Mi ± 1%  -66.62% (p=0.000 n=10)
Exporter/traces/otel/small_batch-10     80.74Ki ± 0%   16.27Ki ± 0%  -79.85% (p=0.000 n=10)
Exporter/traces/otel/medium_batch-10    794.6Ki ± 0%   151.4Ki ± 0%  -80.95% (p=0.000 n=10)
Exporter/traces/otel/large_batch-10     7.739Mi ± 0%   1.462Mi ± 0%  -81.11% (p=0.000 n=10)
Exporter/traces/otel/xlarge_batch-10    77.34Mi ± 0%   14.56Mi ± 0%  -81.18% (p=0.000 n=10)
geomean                                 4.219Mi        967.0Ki       -77.62%

                                      │   old.txt   │               new.txt               │
                                      │  allocs/op  │  allocs/op   vs base                │
Exporter/logs/otel/small_batch-10        553.0 ± 0%    172.0 ± 1%  -68.90% (p=0.000 n=10)
Exporter/logs/otel/medium_batch-10      5.430k ± 0%   1.619k ± 0%  -70.18% (p=0.000 n=10)
Exporter/logs/otel/large_batch-10       54.19k ± 0%   16.08k ± 0%  -70.32% (p=0.000 n=10)
Exporter/logs/otel/xlarge_batch-10      541.7k ± 0%   160.7k ± 0%  -70.34% (p=0.000 n=10)
Exporter/metrics/otel/small_batch-10    4.301k ± 0%   1.918k ± 0%  -55.41% (p=0.000 n=10)
Exporter/metrics/otel/medium_batch-10   42.83k ± 0%   19.00k ± 0%  -55.65% (p=0.000 n=10)
Exporter/metrics/otel/large_batch-10    428.0k ± 0%   189.8k ± 0%  -55.65% (p=0.000 n=10)
Exporter/metrics/otel/xlarge_batch-10   4.278M ± 0%   1.943M ± 0%  -54.59% (p=0.000 n=10)
Exporter/traces/otel/small_batch-10      594.0 ± 0%    192.0 ± 0%  -67.68% (p=0.000 n=10)
Exporter/traces/otel/medium_batch-10    5.830k ± 0%   1.818k ± 0%  -68.82% (p=0.000 n=10)
Exporter/traces/otel/large_batch-10     58.19k ± 0%   18.08k ± 0%  -68.94% (p=0.000 n=10)
Exporter/traces/otel/xlarge_batch-10    581.7k ± 0%   180.6k ± 0%  -68.95% (p=0.000 n=10)
geomean                                 35.09k        12.21k       -65.19%
```

felixbarny commented on exporter/elasticsearchexporter/exporter.go:
```diff
@@ -193,21 +191,18 @@ func (e *elasticsearchExporter) pushMetricsData(
+			var validationErrs []error // log instead of returning these so that upstream does not retry
 			scopeMetrics := scopeMetrics.At(j)
 			scope := scopeMetrics.Scope()
+			groupedDataPointsByIndex := make(map[string]map[uint32][]dataPoint)
```
Note to reviewer: I made it so that documents from different scopes are never merged. This simplified the serialization logic and also fixes a subtle bug in the current implementation where we're only hashing the scope attributes but not the scope name. This leads to grouping of potentially different scopes to the same document. I guess as a consequence, we should also add the scope name as a dimension in the mappings.

axw replied:
I think by moving this here, rather than outside of the `scopeMetrics` loop, we're assuming that there will never be two identical scopes within a resource. Is that a safe assumption?

axw replied:
I suppose it's no worse than the existing assumption that resourceMetrics is free of duplicate resources.

felixbarny replied:
What will make it safe is that the only consequence of being wrong in the assumption is leaving some storage savings on the table. In other words, we should prioritize https://github.com/elastic/elasticsearch/issues/99123, which turns out to be more of an issue than anticipated in various contexts.

axw replied:
Wouldn't duplicates resources/scopes lead to duplicate _tsid & doc rejections?

felixbarny replied:
Yes, it would, until we fix the referenced issue.

felixbarny replied:
Fixing the underlying ES issue turns out to be very tricky. I've created a PR for the ES exporter to enable the grouping across resources and scopes, which hopefully lessens the impact of the bug: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37509.

felixbarny committed:
Merge remote-tracking branch 'origin/main' into es-direct-serialization

felixbarny committed:
Add event_name for logs

felixbarny committed:
Merge branch 'main' into es-direct-serialization

felixbarny committed:
Remove all error handling from serialization code

bytes.Buffer.Write is guaranteed to not return an error

axw commented on exporter/elasticsearchexporter/pdata_serializer.go:
```diff
@@ -0,0 +1,372 @@
+	if attributes.Len() == 0 {
+		return
+	}
+	attrCopy := pcommon.NewMap()
+	attributes.CopyTo(attrCopy)
```
It would be great if we could avoid this copy. The `geo.location` merge might be a bit complicated, but the data stream fields should be reasonably straightforward to filter out by iterating through. A job for another day.

felixbarny replied:
Done in 539fa9d

felixbarny replied:
Relatively speaking, it didn’t reduce allocations a lot. I think the biggest gain would be pooling the buffers. Does docappender support that already?

carsonip replied:
There's a buffer per bulkIndexer, but there is no `sync.Pool`.

axw commented on exporter/elasticsearchexporter/pdata_serializer.go:
```diff
@@ -0,0 +1,372 @@
+	// Determine if this log record is an event, as they are mapped differently
+	// https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md
+	var bodyType string
+	if _, hasEventNameAttribute := record.Attributes().Get("event.name"); hasEventNameAttribute || record.EventName() != "" {
```
We already do this check in the caller, and in a more efficient way, so should we make it a bool parameter?

felixbarny replied:
Done in 7ba257562886184311eee8058ccff4a1c8f485d0

axw reviewed (approved):
LGTM. The amount of handwritten serialisation makes me a little uncomfortable, but we can perhaps improve that with code generation later.

felixbarny committed:
Avoid copying attributes

felixbarny committed:
Propagate isEvent flag to writeLogBody function

felixbarny committed:
Merge remote-tracking branch 'origin/main' into es-direct-serialization

felixbarny committed:
write geo attribute keys

felixbarny committed:
Pool buffers

felixbarny commented:
I've introduced pooling for the buffer in 5e523c5a19966ab3f09c3a087dfb6f6ad40f21a3. This reduced allocation by another 60% and increased throughput as well. I suppose most of the remaining allocations are from creating the pdata model itself, something that we can't easily optimize and not allocations that are directly caused by the ES exporter itself. I've updated the benchmark results in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37032#issuecomment-2577450046.

felixbarny committed:
Add subtext to changelog

felixbarny committed:
Fix checkapi error

exporter/elasticsearchexporter has more than one function: "NewBufferPool,NewFactory"

felixbarny committed:
Merge branch 'main' into es-direct-serialization

felixbarny committed:
gotidy

carsonip commented on exporter/elasticsearchexporter/bufferpol.go:
```diff
@@ -0,0 +1,40 @@
+
+func (p pooledBuffer) WriteTo(w io.Writer) (n int64, err error) {
+	defer p.recycle()
+	return bytes.NewReader(p.Buffer.Bytes()).WriteTo(w)
```
```suggestion
	return p.Buffer.WriteTo(w)
```

nit: any reason to not use WriteTo directly?

carsonip commented on exporter/elasticsearchexporter/bufferpol.go:
```diff
@@ -0,0 +1,40 @@
+// Copyright The OpenTelemetry Authors
```
typo in file name

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -1015,7 +787,7 @@ func sliceHash(h hash.Hash, s pcommon.Slice) {
 // mergeGeolocation mutates attributes map to merge all `geo.location.{lon,lat}`,
```
stale comment

carsonip commented on exporter/elasticsearchexporter/pdata_serializer.go:
```diff
@@ -0,0 +1,381 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package elasticsearchexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter"
```
nit: Seems that this serializer is a good candidate to be moved to a separate package, and only expose the `serialize*` funcs.

felixbarny replied:
I've tried that in the beginning but there are a few package-private things it needs to access. For example, `dataPoint` and `dataStream*` constants in `attribute.go`, which we'd need to make public.

felixbarny replied:
I've moved bufferpool into a separate package now

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -111,24 +112,22 @@ const (
 	default:
 		document = m.encodeLogDefaultMode(resource, record, scope)
 	}
 	// For OTel mode, prefix conflicts are not a problem as otel-data has subobjects: false
 	document.Dedup(m.mode != MappingOTel)
```
```suggestion
	document.Dedup(true)
```
nit: stale code

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -590,128 +480,24 @@ func (dp numberDataPoint) DocCount() uint64 {
 	default:
 		document = m.encodeSpanDefaultMode(resource, span, scope)
 	}
 	// For OTel mode, prefix conflicts are not a problem as otel-data has subobjects: false
 	document.Dedup(m.mode != MappingOTel)
```
```suggestion
	document.Dedup(true)
```
nit: stale code

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -590,128 +480,24 @@ func (dp numberDataPoint) DocCount() uint64 {
-	m.encodeScopeOTelMode(&document, scope, scopeSchemaURL)
-
-	return document
+	err := document.Serialize(buf, m.dedot, m.mode == MappingOTel)
```
```suggestion
	err := document.Serialize(buf, m.dedot, false)
```

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -283,110 +203,65 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
 
-func (m *encodeModel) encodeDocument(document objmodel.Document) ([]byte, error) {
+func (m *encodeModel) encodeDocument(document objmodel.Document, buf *bytes.Buffer) error {
 	// For OTel mode, prefix conflicts are not a problem as otel-data has subobjects: false
 	document.Dedup(m.mode != MappingOTel)
```
```suggestion
	document.Dedup(true)
```

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -283,110 +203,65 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
 
-	var buf bytes.Buffer
-	err := document.Serialize(&buf, m.dedot, m.mode == MappingOTel)
+	err := document.Serialize(buf, m.dedot, m.mode == MappingOTel)
```
```suggestion
	err := document.Serialize(buf, m.dedot, false)
```

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -111,24 +112,22 @@ const (
-	var buf bytes.Buffer
-	err := document.Serialize(&buf, m.dedot, m.mode == MappingOTel)
-	return buf.Bytes(), err
+	return document.Serialize(buf, m.dedot, m.mode == MappingOTel)
```
```suggestion
	return document.Serialize(buf, m.dedot, false)
```

carsonip reviewed (commented):
Thanks, looks good, just a few nits! This will allow us to remove a lot of workarounds due to OTel mode.

felixbarny committed:
Apply suggestions from code review

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

felixbarny committed:
fix stale comment

felixbarny committed:
fix typo in file name

felixbarny committed:
Remove otel serialization code from objmodel

felixbarny committed:
Move bufferpool to dedicated package

felixbarny committed:
Fix geo serialization

felixbarny committed:
Move mergeGeoLocation to pdata_serializer.go

felixbarny committed:
Fix imports

felixbarny committed:
Remove appendValueOnConflict parameter as it's always true

felixbarny committed:
Log validation error when metric with same name has already been serialized

felixbarny committed:
make goporto

felixbarny committed:
Optimize and fix geo attribute serialization

felixbarny committed:
Optimize timestamp serialization

felixbarny committed:
Add todo for more optimization for metrics

felixbarny commented:
4277f76 and 2328a7a add some more minor optimizations that have an average change to the previous commit of +9% events/s and -6.75% B/op.

I have another change lined up to mark the component with `consumer.Capabilities{MutatesData: false}`, which avoids the collector having to clone all input data before calling the exporter. But I'd like to get this PR in first, as it's already quite large.

felixbarny committed:
Merge branch 'main' into es-direct-serialization

felixbarny committed:
Recycle buffer when encoding returns an error

felixbarny committed:
Merge remote-tracking branch 'origin/main' into es-direct-serialization

carsonip commented on exporter/elasticsearchexporter/pdata_serializer.go:
```diff
@@ -331,7 +434,10 @@ func writeValue(v *json.Visitor, val pcommon.Value, stringifyMaps bool) {
 func writeTimestampField(v *json.Visitor, key string, timestamp pcommon.Timestamp) {
 	_ = v.OnKey(key)
-	_ = v.OnString(timestamp.AsTime().UTC().Format(tsLayout))
+	nsec := uint64(timestamp)
```
nit: changing timestamp to a float may make it harder to understand the `telemetry::log_request_body` output, but if it gets us better performance in most cases, it is fine by me

felixbarny replied:
That's fair, but I think the performance gains both here and in ES parsing the timestamp, are worth it.

carsonip reviewed (approved):
lgtm, thanks! Perf gain is wonderful.

felixbarny committed:
Merge branch 'main' into es-direct-serialization

ChrsMark commented:
@JaredTan95 @lahsivjar please take a look

felixbarny committed:
Merge branch 'main' into es-direct-serialization

lahsivjar reviewed (approved):
A bit uncomfortable with the pooling but in the current code it looks good to me.

ChrsMark committed:
Merge branch 'main' into es-direct-serialization

felixbarny committed:
Merge branch 'main' into es-direct-serialization

felixbarny committed:
Merge branch 'main' into es-direct-serialization
